controllers/serviceclaim_controller.go

```diff
@@ -47,13 +61,229 @@ type ServiceClaimReconciler struct {
+		if checkSCISubset(sclaim.Spec.ServiceClassIdentity, rs.Spec.ServiceClassIdentity) {
+			// loop over the environments given in the constraints part of the RegisteredService
+			for _, e := range rs.Spec.Constraints.Environments {
+				// check if the EnvironmentTag of ServiceClaim matches tat of the environment constraint
+				if sclaim.Spec.EnvironmentTag == e {
+					// loop over the ServiceEndpointDefinition array part of RegisteredService
+					for _, sed := range rs.Spec.ServiceEndpointDefinition {
+						// check if the value is non-empty
+						if sed.Value != "" {
+							// check if the ServiceEndpointDefinitionKeys part of ServiceClaim has the current
+							// SED name in the RegisteredService
+							if itemContains(sclaim.Spec.ServiceEndpointDefinitionKeys, sed.Name) {
+								secret.StringData[sed.Name] = sed.Value
+								count = count + 1
+							}
+						} else if sed.ValueFromSecret.Key != "" { // check value if the key is non-empty
+							if itemContains(sclaim.Spec.ServiceEndpointDefinitionKeys, sed.Name) {
+								sec := &corev1.Secret{
+									ObjectMeta: metav1.ObjectMeta{
+										Name:      sed.ValueFromSecret.Name,
+										Namespace: req.NamespacedName.Namespace,
+									},
+								}
+								if err := r.Get(ctx, req.NamespacedName, sec); err != nil {
+									l.Info("unable to retrieve Secret", "error", err)
+									return "", client.IgnoreNotFound(err)
+								}
+								secret.StringData[sed.Name] = sec.StringData[sed.ValueFromSecret.Key]
+								count = count + 1
+							}
+						}
+					}
+					rsName = rs.Name
+				}
+			}
+		} else {
+			sclaim.Status.State = "Pending"
+			if err := r.Status().Update(ctx, &sclaim); err != nil {
+				l.Error(err, "unable to update the ServiceClaim", "ServiceClaim", sclaim)
+				return "", err
+			}
+
+			return "", fmt.Errorf("SCI is not matched")
+		}
+	}
+
+	// if the number of SED keys is more than the number of secret data entries
+	// that indicates one or more keys are missing
+	if len(sclaim.Spec.ServiceEndpointDefinitionKeys) > count {
+		sclaim.Status.State = "Pending"
+		if err := r.Status().Update(ctx, &sclaim); err != nil {
+			l.Error(err, "unable to update the ServiceClaim", "ServiceClaim", sclaim)
+			return "", err
+		}
+
+		return "", fmt.Errorf("Key not available in the list of SEDs")
+	}
+
+	// ServiceClassIdentity values are going to override
+	// any values in the secret resource
+	for _, sci := range sclaim.Spec.ServiceClassIdentity {
+		secret.StringData[sci.Name] = sci.Value
+	}
+
+	var cel primazaiov1alpha1.ClusterEnvironmentList
+	if err := r.List(ctx, &cel); err != nil {
+		l.Info("error fetching ClusterEnvironmentList", "error", err)
+		return "", client.IgnoreNotFound(err)
+	}
+
+	// loop over all ClusterEnvironment
+	for _, ce := range cel.Items {
+		// check if the ServiceClaim EnvironmentTag matches the EnvironmentName part of ClusterEnvironment
+		if sclaim.Spec.EnvironmentTag == ce.Spec.EnvironmentName {
```

Comment:
nesting here can be reduced that way
```suggestion
		if sclaim.Spec.EnvironmentTag != ce.Spec.EnvironmentName {
		        continue
		}
```

Reply:
Done